[feature] 캐시 서비스를 구현한다

hellomatia opened the pull request:
## #️⃣ 연관된 이슈

- #89
- #91
- #92
- #93

## 📝 작업 내용

- 테크 스펙 : [Cache 서버 테크 스펙](https://github.com/woowa-techcamp-2024/Team4-BF2/wiki/Cache-%EC%84%9C%EB%B2%84-%ED%85%8C%ED%81%AC-%EC%8A%A4%ED%8E%99)
### 스크린샷 (선택)

## 💬 리뷰 요구사항(선택)


---
### RCA 룰
| 코드 리뷰 멘토가 PR작성자에게 어떤 의도로 전달 되길 바라는지 알파벳으로 표현해주세요!
- r(request change): 꼭 반영
- c(comment): 왠만하면 반영
- a(approve): 반영해도 좋고 넘어가도 좋은 의견

jcw1031 committed:
[config] 레디스 캐시 모듈 생성

- setting.gradle에 레디스 캐시 모듈 정의
- domain:cache-domain-redis 모듈 추가
- Spring Data Redis 의존성 추가

jcw1031 committed:
[config] local profile 레디스 프로퍼티 설정

- redis host, port 설정

hellomatia committed:
[config] 식당 정보 캐시작업을 수행하는 Service 생성

- 리뷰 통계 정보는 ReviewStatisticsCacheService가 담당한다.
- 식당 정보는 RestaurantCacheService가 담당한다.

jcw1031 committed:
Merge pull request #95 from woowa-techcamp-2024/config/90-cache-configuration

[config] 캐시 서버 환경을 구성한다.

jcw1031 committed:
Merge branch 'dev' into feature/93-integrated-cache-service

jcw1031 committed:
Merge branch 'dev' into feature/93-integrated-cache-service

jcw1031 committed:
[refactor] 캐시 서비스 레포지토리 이름 변경

- RestaurantSummaryRepository -> RestaurantReviewStatisticsRepository
- search 모듈의 레포지토리와 빈 이름 충돌

jcw1031 committed:
[build] 가게 노출 서비스에 캐시 서비스 의존성 설정

- exposure-domain 모듈에 캐시 모듈 의존성 추가
- RestaurantExposureListService에 CacheService 빈 주입

jcw1031 committed:
[build] 캐시 모듈 의존성 설정

- Jackson 라이브러리 의존성 추가
- 도메인 모듈 및 이벤트 모듈 의존성 추가

jcw1031 committed:
[refactor] 불필요한 클래스 제거

- 가게 캐시 서비스와 리뷰 통계 캐시 서비스 제거

jcw1031 committed:
[feature] 가게•리뷰 통계 캐시 서비스 구현

- 가게 ID 목록이 전달되면 Redis에서 조회해 반환
- 만약 Redis에 존재하지 않으면 RDB에서 조회해 Redis에 저장한 후 반환
- 리뷰 통계 변경 이벤트가 발생하면 Redis에 이미 저장된 가게의 리뷰 통계 정보라면 변경

jcw1031 committed:
Merge pull request #99 from woowa-techcamp-2024/feature/93-integrated-cache-service

[feature] 캐시 서비스를 구현한다.

huiseung commented on domain/cache-domain-redis/src/main/java/woowa/team4/bff/cache/redis/service/CacheRedisService.java:
```diff
@@ -0,0 +1,80 @@
+            throw new JsonConvertException(e);
+        }
+    }
+
```
r: @Async 를 하지 않으면 이벤트 발행자가 Listen 처리를 기다리게 돼요. @TransactionEventListener 가 트랜잭션 commit 이후 이벤트를 듣도록 phase 추가를 해주셨으면 해요.

hellomatia committed:
[refactor] JPA DDL 옵션 변경

- 테이블 변경을 방지하기 위해 update -> none으로 변경한다.

hek-git commented on domain/cache-domain-redis/src/main/java/woowa/team4/bff/cache/redis/service/CacheRedisService.java:
```diff
@@ -0,0 +1,80 @@
+        }
+    }
+
+    @TransactionalEventListener
```
a : phase를 아무것도 설정하지 않으면 어떤 시점에 event를 받게 되나요? reviewStatistics가 커밋된 이후에 이벤트를 받지 않는 이유가 궁금합니다!

huiseung commented on mono/src/main/resources/application.yml:
```diff
@@ -19,16 +19,19 @@ spring:
+        enabled: true
+    redis:
+      host: localhost
+      port: 6379
```
a: dev, prod 환경에선 비밀번호 설정 잊지 않아주셨으면 합니다.

huiseung reviewed (changes requested):
수고 하셨습니다. CacheRedisService 수정 부탁드립니다.

hellomatia committed:
[feature] 이벤트 발행자가 Listen처리를 기다리지 않도록 하는 기능 추가

- @Async어노테이션을 추가한다.
- TransactionEventListener가 트랜잭션 commit이후에 이벤트를 듣도록 phase를 추가한다.

hellomatia committed:
[fix] 리뷰 생성시 이벤트를 중복적으로 발행하는 버그 해결

- 중복으로 이벤트를 발행하여 무한으로 updateReviewStatistics를 실행시키는 현상을 수정한다.

hek-git reviewed (approved):
캐싱 구현 수고하셨습니다!

huiseung reviewed (approved):
수고하셨습니다.
